sw/device/silicon_creator/lib/drivers/ast.c

```diff
@@ -0,0 +1,77 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "sw/device/silicon_creator/lib/drivers/ast.h"
+
+#include "sw/device/lib/base/abs_mmio.h"
+#include "sw/device/lib/base/multibits.h"
+#include "sw/device/silicon_creator/lib/drivers/otp.h"
+
+#include "hw/top_earlgrey/sw/autogen/top_earlgrey.h"
+#include "otp_ctrl_regs.h"     // Generated.
+#include "sensor_ctrl_regs.h"  // Generated.
+
+enum {
+  kSensorCtrlBase = TOP_EARLGREY_SENSOR_CTRL_BASE_ADDR,
+};
+
+rom_error_t ast_check(lifecycle_state_t lc_state) {
+  // OTP can be configured to skip AST initialization. In this situation we do
+  // not check that AST_INIT_DONE is set.
+  uint32_t en = otp_read32(OTP_CTRL_PARAM_CREATOR_SW_CFG_AST_INIT_EN_OFFSET);
+  if (launder32(en) == kMultiBitBool4False) {
+    HARDENED_CHECK_EQ(en, kMultiBitBool4False);
+    return kErrorOk;
+  }
+
+  // In some lifecycle states we want to continue the boot process even if the
+  // AST is not initialized. Note that in these states OTP may not have been
+  // configured.
+  switch (launder32(lc_state)) {
+    case kLcStateTest:
+      HARDENED_CHECK_EQ(lc_state, kLcStateTest);
+      return kErrorOk;
+    case kLcStateRma:
+      HARDENED_CHECK_EQ(lc_state, kLcStateRma);
+      return kErrorOk;
+    case kLcStateDev:
+      HARDENED_CHECK_EQ(lc_state, kLcStateDev);
+      break;
+    case kLcStateProd:
+      HARDENED_CHECK_EQ(lc_state, kLcStateProd);
+      break;
+    case kLcStateProdEnd:
+      HARDENED_CHECK_EQ(lc_state, kLcStateProdEnd);
+      break;
+    default:
+      return kErrorUnknown;
+  }
+
+  // If OTP has not been configured correctly then the AST configuration may
+  // have been skipped. Report a specific error in this scenario.
+  if (en != kMultiBitBool4True) {
+    return kErrorAstBadConfiguration;
+  }
```

Comment:
This is just a debugging aid really. If we remove this check and `AST_INIT_EN` in OTP is set to something other than `kMultiBitBool4True` or `kMultiBitBool4False` then AST *will not* be initialized but the init done check *will* still happen. This check means that a specific error is generated instead of the init done check just timing out.

Reply:
I removed this part since we should cover this in a test.